cmd/rpcdaemon/commands/eth_filters.go

```diff
@@ -13,29 +17,129 @@ import (
 )
 
 // NewPendingTransactionFilter new transaction filter
-func (api *APIImpl) NewPendingTransactionFilter(ctx context.Context) (hexutil.Uint64, error) {
-	return 0, fmt.Errorf(NotImplemented, "eth_newPendingTransactionFilter")
+func (api *APIImpl) NewPendingTransactionFilter(_ context.Context) (common.Hash, error) {
+	if api.filters == nil {
+		return common.Hash{}, rpc.ErrNotificationsUnsupported
+	}
+	txsCh := make(chan []types.Transaction, 1)
+	id := api.filters.SubscribePendingTxs(txsCh)
+	go func() {
+		for {
+			select {
+			case txs, ok := <-txsCh:
+				if !ok {
+					return
+				}
+				api.filters.AddPendingTxs(id, txs)
+			default:
+			}
+		}
+	}()
+	return common.HexToHash(string(id)), nil
 }
 
-// NewBlockFilter new transaction filter
-func (api *APIImpl) NewBlockFilter(_ context.Context) (hexutil.Uint64, error) {
-	return 0, fmt.Errorf(NotImplemented, "eth_newBlockFilter")
+// NewBlockFilter implements eth_newBlockFilter. Creates a filter in the node, to notify when a new block arrives.
+func (api *APIImpl) NewBlockFilter(_ context.Context) (common.Hash, error) {
+	if api.filters == nil {
+		return common.Hash{}, rpc.ErrNotificationsUnsupported
+	}
+	ch := make(chan *types.Block, 1)
+	id := api.filters.SubscribePendingBlock(ch)
+	go func() {
+		for {
+			select {
+			case block, ok := <-ch:
+				if !ok {
+					return
+				}
+				api.filters.AddPendingBlock(id, block)
+			default:
+			}
+		}
+	}()
+	return common.HexToHash(string(id)), nil
 }
 
 // NewFilter implements eth_newFilter. Creates an arbitrary filter object, based on filter options, to notify when the state changes (logs).
-func (api *APIImpl) NewFilter(_ context.Context, filter interface{}) (hexutil.Uint64, error) {
-	return 0, fmt.Errorf(NotImplemented, "eth_newFilter")
+func (api *APIImpl) NewFilter(_ context.Context, crit filters.FilterCriteria) (common.Hash, error) {
+	if api.filters == nil {
+		return common.Hash{}, rpc.ErrNotificationsUnsupported
+	}
+	logs := make(chan *types.Log, 1)
+	id := api.filters.SubscribeLogs(logs, crit)
+	go func() {
+		for {
+			select {
+			case log, ok := <-logs:
```

Comment:
Add ctx.Done() to select

Reply:
Will it not be Done() right after the end of the request?

Reply:
"end of the request" will happen on next line `return common.HexToHash`, then `UninstallFilter` it's already another request. What if client crush and never call `UninstallFilter`? will channel `logs` and this goroutine will leak in this case?

Reply:
This "newFilter" stuff is supposed to gather information between getFilterChanges requests. So if we close the goroutine at ctx.Done() it will gather nothing.
We've discussed it with @AlexeyAkhunov today. He approved that these functions are meant to be like this, but must be restricted for users (only trusted users can use them).

Reply:
"So if we close the goroutine at ctx.Done() it will gather nothing." - agree

Reply:
ok. then go-go! likely this funks will not have high RPS